File: examples/go/merkle-proof/MerkleProofDemo_test.go
package contract

import (
	"testing"

	runar "github.com/icellan/runar/packages/runar-go"
)

// ---------------------------------------------------------------------------
// Merkle tree helper — builds a depth-4 tree (16 leaves)
// ---------------------------------------------------------------------------

type merkleTree struct {
	root   runar.ByteString
	leaves []runar.ByteString
	layers [][]runar.ByteString
}

func buildMerkleTree(leaves []runar.ByteString, hashFn func(runar.ByteString) runar.ByteString) merkleTree {
	level := make([]runar.ByteString, len(leaves))
	copy(level, leaves)
	layers := [][]runar.ByteString{level}

	for len(level) > 1 {
		next := make([]runar.ByteString, 0, len(level)/2)
		for i := 0; i < len(level); i += 2 {
			next = append(next, hashFn(level[i]+level[i+1]))
		}
		level = next
		layers = append(layers, level)
	}

	return merkleTree{
		root:   level[0],
		leaves: leaves,
		layers: layers,
	}
}

func (t *merkleTree) getProof(index int) (leaf, proof runar.ByteString) {
	var siblings []runar.ByteString
	idx := index
	for d := 0; d < len(t.layers)-1; d++ {
		siblingIdx := idx ^ 1
		siblings = append(siblings, t.layers[d][siblingIdx])
		idx >>= 1
	}
	var proofConcat runar.ByteString
	for _, s := range siblings {
		proofConcat += s
	}
	return t.leaves[index], proofConcat
}

// Create 16 leaves (32-byte SHA-256 hashes)
func makeLeaves() []runar.ByteString {
	leaves := make([]runar.ByteString, 16)
	for i := 0; i < 16; i++ {
		leaves[i] = runar.Sha256Hash(runar.ByteString([]byte{byte(i)}))
	}
	return leaves
}

// ---------------------------------------------------------------------------
// verifySha256 (merkleRootSha256, depth=4)
// ---------------------------------------------------------------------------

func TestMerkleProofDemo_VerifySha256_Index0(t *testing.T) {
	leaves := makeLeaves()
	tree := buildMerkleTree(leaves, runar.Sha256Hash)
	leaf, proof := tree.getProof(0)

	c := &MerkleProofDemo{ExpectedRoot: tree.root}
	c.VerifySha256(leaf, proof, 0)
}

func TestMerkleProofDemo_VerifySha256_Index7(t *testing.T) {
	leaves := makeLeaves()
	tree := buildMerkleTree(leaves, runar.Sha256Hash)
	leaf, proof := tree.getProof(7)

	c := &MerkleProofDemo{ExpectedRoot: tree.root}
	c.VerifySha256(leaf, proof, 7)
}

func TestMerkleProofDemo_VerifySha256_Index15(t *testing.T) {
	leaves := makeLeaves()
	tree := buildMerkleTree(leaves, runar.Sha256Hash)
	leaf, proof := tree.getProof(15)

	c := &MerkleProofDemo{ExpectedRoot: tree.root}
	c.VerifySha256(leaf, proof, 15)
}

func TestMerkleProofDemo_VerifySha256_WrongLeaf(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected assertion failure for wrong leaf")
		}
	}()

	leaves := makeLeaves()
	tree := buildMerkleTree(leaves, runar.Sha256Hash)
	_, proof := tree.getProof(0)
	wrongLeaf := runar.Sha256Hash(runar.ByteString([]byte{0xff}))

	c := &MerkleProofDemo{ExpectedRoot: tree.root}
	c.VerifySha256(wrongLeaf, proof, 0)
}

func TestMerkleProofDemo_VerifySha256_WrongIndex(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected assertion failure for wrong index")
		}
	}()

	leaves := makeLeaves()
	tree := buildMerkleTree(leaves, runar.Sha256Hash)
	leaf, proof := tree.getProof(0)

	c := &MerkleProofDemo{ExpectedRoot: tree.root}
	c.VerifySha256(leaf, proof, 1) // wrong index
}

// ---------------------------------------------------------------------------
// verifyHash256 (merkleRootHash256, depth=4)
// ---------------------------------------------------------------------------

func TestMerkleProofDemo_VerifyHash256_Index0(t *testing.T) {
	leaves := makeLeaves()
	tree := buildMerkleTree(leaves, runar.Hash256)
	leaf, proof := tree.getProof(0)

	c := &MerkleProofDemo{ExpectedRoot: tree.root}
	c.VerifyHash256(leaf, proof, 0)
}

func TestMerkleProofDemo_VerifyHash256_Index10(t *testing.T) {
	leaves := makeLeaves()
	tree := buildMerkleTree(leaves, runar.Hash256)
	leaf, proof := tree.getProof(10)

	c := &MerkleProofDemo{ExpectedRoot: tree.root}
	c.VerifyHash256(leaf, proof, 10)
}

// ---------------------------------------------------------------------------
// Compile check
// ---------------------------------------------------------------------------

func TestMerkleProofDemo_Compile(t *testing.T) {
	if err := runar.CompileCheck("MerkleProofDemo.runar.go"); err != nil {
		t.Fatalf("Runar compile check failed: %v", err)
	}
}

File: examples/rust/merkle-proof/MerkleProofDemo_test.rs
#[path = "MerkleProofDemo.runar.rs"]
mod contract;

use contract::*;
use runar::prelude::*;

// ---------------------------------------------------------------------------
// Merkle tree helper -- builds a depth-4 tree (16 leaves)
// ---------------------------------------------------------------------------

struct MerkleTree {
    root: ByteString,
    leaves: Vec<ByteString>,
    layers: Vec<Vec<ByteString>>,
}

fn build_merkle_tree(
    leaves: &[ByteString],
    hash_fn: fn(&[u8]) -> ByteString,
) -> MerkleTree {
    let mut level: Vec<ByteString> = leaves.to_vec();
    let mut layers: Vec<Vec<ByteString>> = vec![level.clone()];

    while level.len() > 1 {
        let mut next = Vec::new();
        for i in (0..level.len()).step_by(2) {
            let mut preimage = level[i].clone();
            preimage.extend_from_slice(&level[i + 1]);
            next.push(hash_fn(&preimage));
        }
        level = next;
        layers.push(level.clone());
    }

    MerkleTree {
        root: level[0].clone(),
        leaves: leaves.to_vec(),
        layers,
    }
}

fn get_proof(tree: &MerkleTree, index: usize) -> (ByteString, ByteString) {
    let mut siblings = Vec::new();
    let mut idx = index;
    for d in 0..tree.layers.len() - 1 {
        let sibling_idx = idx ^ 1;
        siblings.push(tree.layers[d][sibling_idx].clone());
        idx >>= 1;
    }
    let mut proof = Vec::new();
    for s in &siblings {
        proof.extend_from_slice(s);
    }
    (tree.leaves[index].clone(), proof)
}

fn make_leaves() -> Vec<ByteString> {
    (0u8..16).map(|i| sha256(&[i])).collect()
}

// ---------------------------------------------------------------------------
// verify_sha256 (merkle_root_sha256, depth=4)
// ---------------------------------------------------------------------------

#[test]
fn test_verify_sha256_index0() {
    let leaves = make_leaves();
    let tree = build_merkle_tree(&leaves, sha256);
    let (leaf, proof) = get_proof(&tree, 0);

    let c = MerkleProofDemo { expected_root: tree.root };
    c.verify_sha256(leaf, proof, 0);
}

#[test]
fn test_verify_sha256_index7() {
    let leaves = make_leaves();
    let tree = build_merkle_tree(&leaves, sha256);
    let (leaf, proof) = get_proof(&tree, 7);

    let c = MerkleProofDemo { expected_root: tree.root };
    c.verify_sha256(leaf, proof, 7);
}

#[test]
fn test_verify_sha256_index15() {
    let leaves = make_leaves();
    let tree = build_merkle_tree(&leaves, sha256);
    let (leaf, proof) = get_proof(&tree, 15);

    let c = MerkleProofDemo { expected_root: tree.root };
    c.verify_sha256(leaf, proof, 15);
}

#[test]
#[should_panic]
fn test_verify_sha256_wrong_leaf() {
    let leaves = make_leaves();
    let tree = build_merkle_tree(&leaves, sha256);
    let (_, proof) = get_proof(&tree, 0);
    let wrong_leaf = sha256(&[0xff]);

    let c = MerkleProofDemo { expected_root: tree.root };
    c.verify_sha256(wrong_leaf, proof, 0);
}

#[test]
#[should_panic]
fn test_verify_sha256_wrong_index() {
    let leaves = make_leaves();
    let tree = build_merkle_tree(&leaves, sha256);
    let (leaf, proof) = get_proof(&tree, 0);

    let c = MerkleProofDemo { expected_root: tree.root };
    c.verify_sha256(leaf, proof, 1); // wrong index
}

// ---------------------------------------------------------------------------
// verify_hash256 (merkle_root_hash256, depth=4)
// ---------------------------------------------------------------------------

#[test]
fn test_verify_hash256_index0() {
    let leaves = make_leaves();
    let tree = build_merkle_tree(&leaves, hash256);
    let (leaf, proof) = get_proof(&tree, 0);

    let c = MerkleProofDemo { expected_root: tree.root };
    c.verify_hash256(leaf, proof, 0);
}

#[test]
fn test_verify_hash256_index10() {
    let leaves = make_leaves();
    let tree = build_merkle_tree(&leaves, hash256);
    let (leaf, proof) = get_proof(&tree, 10);

    let c = MerkleProofDemo { expected_root: tree.root };
    c.verify_hash256(leaf, proof, 10);
}

// ---------------------------------------------------------------------------
// Compile check
// ---------------------------------------------------------------------------

#[test]
fn test_compile() {
    runar::compile_check(
        include_str!("MerkleProofDemo.runar.rs"),
        "MerkleProofDemo.runar.rs",
    )
    .unwrap();
}
